packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewer/useAsyncFunctionCall.ts

```diff
@@ -0,0 +1,46 @@
+import { useEffect, useState } from 'react';
+
+type AsyncFunc<ReturnType> = (signal: AbortSignal) => Promise<ReturnType>;
+type AsyncFuncReturnType<T> = T extends AsyncFunc<infer U> ? U : never;
+
+/**
+ * Call async function WRAPPED BY `useCallback` and return its result
+ *
+ * @param asyncFunction async function wrapped by `useCallback`.
+ *  Take one parameter `setCancellable` to set _cancellable_ of the current async call.
+ * @returns the result of the async function
+ */
+function useAsyncFunctionCall<Func extends AsyncFunc<any>, ReturnType = AsyncFuncReturnType<Func>>(
+  asyncFunction: Func
+): ReturnType | undefined {
+  const [result, setResult] = useState<ReturnType | undefined>();
+
+  useEffect(() => {
+    let state: 'pending' | 'fulfilled' | 'rejected' = 'pending';
```

Comment:
Nit: seems this function only needs to differentiate between `pending` or not, so this could be simplified to just a simple boolean.